BlobInputStream improvements

OrangeDog opened the pull request:
* fix: inconsistent behaviour of BlobInputStream.reset()

  A BlobInputStream will start reading from the current position of the LargeObject used to create it, but the current/marked position does not account for this.
Initialising current position correctly will fix this.

  Closes #3149
* feat: implement efficient skipping in BlobInputStream

  Using seek is more efficient than reading and discarding the bytes. The only behavioural change is it now allows skipping  past the natural end of the stream, but the method contract already allows for this. It should not break anything.

  Closes #3144

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Does `./gradlew styleCheck` pass ?
3. [x] Have you added your new test classes to an existing test suite ~in alphabetical order~?

### Changes to Existing Features:

* [x] Does this break existing behaviour? If so please explain.
* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

OrangeDog commented:
Both can be solved in the same way!

In the constructor call `tell64()`, then if it throws call `tell()`.
Then you know both the current position, and the max blob length.

OrangeDog committed:
fix: inconsistent behaviour of BlobInputStream.reset()

A BlobInputStream will start reading from the current position of the LargeObject used to create it, but the current/marked position does not account for this.
Initialising current position correctly will fix this.

Closes #3149

OrangeDog committed:
feat: implement efficient skipping in BlobInputStream

Using seek is more efficient than reading and discarding the bytes.
The only behavioural change is it now allows skipping past the natural end of the stream, but the method contract already allows for this. It should not break anything.

Closes #3144

davecramer commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -90,8 +90,25 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
+    long maxBlobSize;
+    try {
+      this.absolutePosition = lo.tell64();
+      maxBlobSize = 4L * 1024 * 1024 * 1024 * 1024 - 2 * 1024;
```
Can you provide a link or explanation for this math ? I realize the docs say 4TB. Link to that would be good

OrangeDog replied:
The docs indeed say 4TB, but you cannot actually get that big (with PG14 anyway).
I don't know why the real limit is 2kB less than 4TB.

vlsi replied:
What is the point of having the math in the driver? What if PG 19 lifts the limit to 4 exabytes?

OrangeDog replied:
How do you suggest determining the correct value otherwise?

If the limit is raised to 4 exabytes, then the `LargeObject` API will also have to change to take a value larger than a `long`, so all this code would need changing to support it anyway.

vlsi replied:
Signed Long.MAX_VALUE is 9223372036854775807L (9 exabytes), and unsigned long is ~18 exabytes, so the current `seek64(long pos` is more-or-less enough for multi-exabyte long values. I'm not sure what is the point of artificially limiting the values at the driver level

OrangeDog replied:
> Signed Long.MAX_VALUE is 9223372036854775807L (9 exabytes)

That's 8 EiB, but you're right, I miscounted the digits.

> what is the point of artificially limiting the values at the driver level

If you don't then it throws. If you prefer that behaviour then I can remove `maxBlobSize`, but that would be a breaking change for `skip()` unless we also catch and check the error message.

esabol replied:
> Can you provide a link or explanation for this math ? I realize the docs say 4TB. Link to that would be good

Here is a link that you might be able to use:

https://www.enterprisedb.com/postgres-tutorials/postgresql-toast-and-working-blobsclobs-explained

It seems crazy, but I think the only mention of the 4TB upper limit on blobs on postgresql.org is in the release notes for 9.3 under "E.26.3.5. Data Types": https://www.postgresql.org/docs/9.3/release-9-3.html

OrangeDog replied:
@esabol: https://www.postgresql.org/docs/16/lo-intro.html

The problem is that it's _not_ 4 TB (TiB). It's slightly less.

esabol replied:
> The problem is that it's _not_ 4 TB (TiB). It's slightly less.

I'm aware of that. @davecramer asked for a link (to be added as a comment in the code), so I thought I'd offer what I could find.

OrangeDog replied:
I found it: https://github.com/postgres/postgres/blob/REL_16_2/src/include/storage/large_object.h#L76

That's `(2^31 - 1) * 2048 == 4 * 1024^4 - 2048`

I'll change the code here to match, with a link.

davecramer commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -90,8 +90,25 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
+      maxBlobSize = 4L * 1024 * 1024 * 1024 * 1024 - 2 * 1024;
+    } catch (SQLException e1) {
+      try {
+        this.absolutePosition = lo.tell();
```
Document what this is doing ?

OrangeDog replied:
This line specifically, or the added section?

davecramer replied:
seems to me that we are switching from LO to bytea here ?

vlsi replied:
The whole section. Frankly, the code is non obvious, so it makes sense to document the considered edge cases

vlsi replied:
It is sad we increase the number of roundtrips for a common case of "blob is used only once". Can we avoid adding extra db calls in the constructor?

OrangeDog replied:
No? On old version of Postgres (which you still test against), large objects only use an int for their position, with a limit of 2GB: https://www.postgresql.org/docs/9.2/lo-intro.html

See also the existing implementation of `reset()`, which checks whether it can use `seek` instead of `seek64`.

OrangeDog replied:
> Can we avoid adding extra db calls in the constructor?

Not really. Someone might call `mark()` before anything else, and the position needs to be correctly initialised before then.

davecramer replied:
As far as testing goes, if you don't want to test on older versions just disable the test. I'm fine with that.

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,6 +352,56 @@ public boolean markSupported() {
+          }
+        } catch (SQLException e) {
+          throw new IOException(
+              GT.tr("Can not skip stream for large object {0} by {1}",
```
Wdyt of adding the current position to the error message?

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -200,26 +203,85 @@ void uploadBlob_NATIVE() throws Exception {
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.WRITE)) {
+      byte[] buf = new byte[1024];
+      for (byte i = 0; i < 96; i++) {
+        Arrays.fill(buf, i);
```
I would prefer random data since 11111 22222 333 might mask bugs in case read accesses a wrong buffer item

OrangeDog replied:
The test has to know which value is expected. Random data would be difficult to verify by hand.

As there are only 256 possible byte values, there will always be duplicates in the data, so you cannot prove it wasn't accidentally reading a different byte that happened to have the same value.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -200,26 +203,85 @@ void uploadBlob_NATIVE() throws Exception {
+      assertThrows(IOException.class, () -> bis.skip(1));
+    }
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
```
Can you please use parameterized tests or at least individual test methods with their own names? Currently, it is hard to understand the intention behind the added tests

vlsi reviewed (changes requested):
Looks easier than I thought

OrangeDog commented:
@davecramer  @vlsi can I get your input to finish resolving all the above?

If necessary, I can split the bug fix commit to a separate PR.

OrangeDog commented:
@davecramer @vlsi I've not had any responses for two months now

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -207,21 +210,94 @@ void markResetStream() throws Exception {
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(0, bis.read());
+      assertEquals(1024L, bis.skip(1024));
+      assertEquals(1, bis.read());
+      assertEquals(0L, bis.skip(-1));
+      assertEquals(1, bis.read());
+      assertEquals(64 * 1024L, bis.skip(64 * 1024));
+      assertEquals(65, bis.read());
+      assertEquals(expectedBigSkip, bis.skip(Long.MAX_VALUE));
+      assertEquals(-1, bis.read());
```
Please add the corresponding messages describing the reason why expect certain values. If something goes wrong, it would be hard to tell why 65 was expected at line 278.

OrangeDog replied:
I'm not really sure how to phrase it. What exactly do you want?

The reason why it should be 65 etc.  is obvious if you know the structure of the file. I can add a comment to each test describing it, if the code in `createMediumLargeObject` is insufficient?

Or do you want the aspect of the implementation that is being tested? Is there any preference on phrasing, given that it's supposed to describe the failure, not describe what it was testing?

No other assertions in this project use messages, so I have nothing to go on.

vlsi replied:
> The reason why it should be 65 etc. is obvious if you know the structure of the 

I am afraid it is not obvious.

> No other assertions in this project use messages

I am afraid that is not quite right.
Please check here: https://github.com/pgjdbc/pgjdbc/blob/d12b625059310ce40548ff448f9485f5bd52b035/pgjdbc/src/test/java/org/postgresql/util/LazyCleanerTest.java#L63-L103
https://github.com/pgjdbc/pgjdbc/blob/d12b625059310ce40548ff448f9485f5bd52b035/pgjdbc/src/test/java/org/postgresql/test/jdbc2/BatchedInsertReWriteEnabledTest.java#L421-L433

> Is there any preference on phrasing, given that it's supposed to describe the failure, not describe what it was testing?

https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md#test

> CONTRIBUTING.md#test: General rule: failing test should look like a good bug report

In other words, the message should describe the reason you expect something to equal to a something else.

Note: if the test class and test method gives enough information, it might be fine to omit the message.
However, in your case you put like ~10 assertions one after another, so it is important to make the failures distinct.

---

> No other assertions in this project use messages, so I have nothing to go on

There's maintenance cost. If the test fails sometime in the future, the one who observes the failure would have to pay the cost of analyzing the failure and deciphering the intentions behind the test. Of course the current code is not all perfect, however, it is important to keep the added code maintainable.

vlsi replied:
It might be some of the assertions could be split into their own test methods. Then the failures would be easier to analyze and reproduce.

OrangeDog replied:
> I am afraid it is not obvious.

I don't know how to write a good explanation because it really is obvious to me.

If you know a stream goes `[1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4]`, and you do `read(); skip(5); read()` then how it it not obvious that the results should be `1` and `3`?

I cannot put the full structure of the file into every assert message. I can put things like "failed to skip beyond internal buffer". That's what would be wrong with it. The value returned is arbitrary, but as previously discussed I chose this structure of values so that it's easy to see what the expected value should be, and minimise the chance of a fluke pass.

OrangeDog replied:
> It might be some of the assertions could be split into their own test methods. Then the failures would be easier to analyze and reproduce.

I've done that a little, now that seeking too far throws and breaks the transaction. I was following the style of the existing tests here, which all have multiple assertions. Especially with streams you need multiple previous operations to set up the state you're testing. e.g. having had the buffer reallocate, being in a position where it might go backwards, etc.

I'll push the next updates without messages and more separate methods, and then see what you think.

OrangeDog replied:
@vlsi it is done

OrangeDog commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -90,8 +92,30 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
+      // initialise current position for mark/reset
+      this.absolutePosition = lo.tell64();
+      // https://github.com/postgres/postgres/blob/REL9_3_0/src/include/storage/large_object.h#L76
+      maxBlobSize = (long) Integer.MAX_VALUE * LO_BLOCK_SIZE;
```
Because that is a breaking change to the behaviour of `skip(long)`.
https://github.com/pgjdbc/pgjdbc/pull/3148#discussion_r1513081689

What it currently does is skip to the end of the current input for all values larger than the remaining input. This includes returning 0 if it's already at the end.

I had another idea recently. We could provide a setter to allow users to change it if they have built postgres with a custom block size.

vlsi replied:
I suggest we defer the call to the database. I do not see value in artificially adding a limit on the driver's side.

OrangeDog replied:
So I therefore need to split the PR so that the bug fix doesn't have to wait for v43?

OrangeDog replied:
Turns out older postgres behaves differently, looking at the 8.4 test failure. I will investigate further.

OrangeDog committed:
feat: do not attempt to guard server-side large object limit

BlobInputStream.seek will now throw if attempting to skip beyond the max size for a large object. This is a breaking change. The previous implementation was imperfect as it did not account for the server being compiled with a different limit.

The state of the stream after the exception is undefined, but hopefully sane. Further exceptions may be thrown on future operations.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -290,18 +298,183 @@ void markResetStream() throws Exception {
+
+    try (LargeObject blob = lom.open(loid, LargeObjectManager.READ)) {
+      InputStream bis = blob.getInputStream();
+      assertEquals(0, bis.read());
+      assertEquals(1024L, bis.skip(1024));
+      assertEquals(1, bis.read());
+      assertEquals(0L, bis.skip(-1024));
```
Is read-skip-read needed for "skip backwards" test?
I guess it should be enough to have something like `read(); skip(-1); read()`.

OrangeDog replied:
For this test file it is, because of the structure. I guess I can change it to use the xml file instead, as this test is not exercising the buffering code.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -290,18 +298,183 @@ void markResetStream() throws Exception {
+
+  @Test
+  void skipPastMaxAfter9_3() throws Exception {
+    assumeTrue(TestUtil.haveMinimumServerVersion(con, ServerVersion.v9_3));
```
This could have been `@DisabledIfServerVersionBelow("9.3")`

OrangeDog replied:
Indeed. I didn't know about it. Other tests in this file don't even skip, they just pass.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -290,18 +298,183 @@ void markResetStream() throws Exception {
+            // consume stream
+          }
+        });
+        assertThrows(IOException.class, bis::close);
```
Why is it expected to throw an error? What is the expected error?

OrangeDog replied:
Because the previous server error invalidates the transaction, causing all further operations to throw.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/BlobTest.java:
```diff
@@ -290,18 +298,183 @@ void markResetStream() throws Exception {
-        bis.reset();
-        assertEquals('?', bis.read());
+    assertEquals(PSQLState.IN_FAILED_SQL_TRANSACTION.getState(), ex.getSQLState());
+    assertTrue(completedTest.get());
```
Can you please clarify how this can be `true` along with `PSQLException ex = assertThrows(PSQLException.class, () -> {` above?

Does it mean you expect an error out of `blob.close()`?
If that is the case, I would structure the test explicitly, without try-with-resources, otherwise it is hard to understand what is actually tested here and how it is supposed to work

OrangeDog replied:
Yes, it throws on close. I'll change it as suggested.

OrangeDog commented:
The latest test run had a failure in `org.postgresql.replication.LogicalReplicationStatusTest`. I'm pretty sure that wasn't me.

OrangeDog commented:
It's been months since I last addressed the feedback. Can someone please look at this?

davecramer commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -90,8 +90,24 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
+
+    try {
+      // initialise current position for mark/reset
+      this.absolutePosition = lo.tell64();
```
So we have access to the baseConnection through `lo.fp.connection` we can get the server version from there and avoid this dance.

OrangeDog replied:
We could avoid the try/catch, yes, but we still need to do the `tell` to fix #3149.

davecramer replied:
agreed. As much as I dislike putting the call in the constructor I don't think it makes that much of a difference.

davecramer replied:
After contemplating this some more, I think it makes more sense to put the tell in getLo(). We expect read/write move operations to throw exceptions, we do not expect constructors to .

OrangeDog replied:
But the constructor here is public. This class has to be able to deal with a passed lo that is not at the start.

I don't see why constructors cannot throw, but (as previously discussed) the only other solution is to defer this check to the first time any other method is called, which makes for brittle and repetitive code.

davecramer replied:
It would be very surprising to me if my java program threw a `RuntimeException` because of a connection error. If you put it in getLo() which is called by every read/write/move operation it would only be in one place, there would not be duplicate or repetitive code and we expect a SQLException here .

OrangeDog replied:
Oh I see, sorry I forgot where that method was. OK, I'll look at doing that.

OrangeDog replied:
Sorry, that doesn't work. `mark(int)` is not allowed to throw any checked exceptions, but needs to have had the position initialized.

OrangeDog replied:
Also "we have access to the baseConnection" isn't true. Those fields are private.

davecramer replied:
Looking at mark, we don't even implement that correctly as we ignore the readLimit. I'd be fine with ignoring the exception there. After all you can call mark on a closed input stream.

OrangeDog replied:
There are no issues with the `readLimit`. The mark position is never invalidated by reading. It is there to allow a stream to allocate a buffer when its underlying source cannot seek backwards.

Yes, ignoring exceptions there should be OK, though I'll still log it so that the cause of the next exception is not completely lost. I should have that done tomorrow.

vlsi replied:
I do not like the idea of throwing an exception on every `lo.tell64` call.

I would rather suggest implementing `int FastPath#getServerVersionNum()`, so `LargeObject#tell64` could automatically call `lo_tell` if the database does not support tell64.

OrangeDog replied:
That sounds terrible. `LargeObject#tell64` should always call `tell64` and `LargeObject#tell` should always call `tell`.

vlsi replied:
Please come up with the PR that fulfills the following then:
* it should work with the supported DB versions
* it should not degrade performance
* it should support LO exceeding MAX_INT

OrangeDog replied:
That is what this PR does...

vlsi replied:
I'm afraid the PR "degrades the performance for PG<9.3" since it would always try `tell64`, catch the exception and try `tell`. Would you please be so kind to revalidate and ensure the PR fulfills the above requirements?

OrangeDog replied:
This PR massively improves the performance for all postgres versions by allowing it to actually seek instead of having to fetch all the data it's trying to skip over.

The single additional call I have added for the bug fix is unavoidable if you want the behaviour to be correct. The options are always two call (check version, then use correct tell), or only one call for the majority of users (try tell64, then tell if it fails).

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,6 +351,59 @@ public boolean markSupported() {
+      if (buffer != null && buffer.length - bufferPosition > skipped) {
+        bufferPosition += (int) skipped;
+      } else {
+        try {
+          if (targetPosition <= Integer.MAX_VALUE) {
+            lo.seek((int) targetPosition, LargeObject.SEEK_SET);
+          } else {
+            lo.seek64(targetPosition, LargeObject.SEEK_SET);
+          }
+        } catch (SQLException e) {
+          throw new IOException(
+              GT.tr("Can not skip stream for large object {0} by {1} (currently @{2})",
+                  loId, n, currentPosition),
+              e);
+        }
+        buffer = null;
```
```suggestion
        buffer = null;
        try {
          if (targetPosition <= Integer.MAX_VALUE) {
            lo.seek((int) targetPosition, LargeObject.SEEK_SET);
          } else {
            lo.seek64(targetPosition, LargeObject.SEEK_SET);
          }
        } catch (SQLException e) {
          throw new IOException(
              GT.tr("Can not skip stream for large object {0} by {1} (currently @{2})",
                  loId, n, currentPosition),
              e);
        }
```

The `buffer` should be invalidated at the very beginning to avoid stale data in case of failure.

OrangeDog replied:
If any of these methods throw then you don't need to guarentee what continuing to use the stream will do, but sure.

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,6 +351,59 @@ public boolean markSupported() {
+        }
+        buffer = null;
+      }
+      absolutePosition = targetPosition;
```
In fact, `lo_seek` returns the new location pointer, so I think we should use it directly rather than assuming `targetPosition`.

OrangeDog replied:
However, `LargeObject#seek` does not.

OrangeDog replied:
That could be a future job, yes. I haven't changed the current way of doing it, and it works just fine right now. Can we not scope-creep this PR to find more things to argue about?

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,6 +351,59 @@ public boolean markSupported() {
+      } else {
+        try {
+          if (targetPosition <= Integer.MAX_VALUE) {
+            lo.seek((int) targetPosition, LargeObject.SEEK_SET);
```
See https://www.postgresql.org/docs/15/lo-interfaces.html#LO-SEEK

> Note that lo_lseek will fail if the new location pointer would be greater than 2GB.

I suggest we should always use `seek64` here, and `LargeObject` should use either 64 or and old api depending on the backend version.

OrangeDog replied:
The previous developer already accounted for that. That's why it's using `SEEK_SET` and not `SEEK_CUR`.

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,6 +351,59 @@ public boolean markSupported() {
+  /**
+   * Skips over and discards {@code n} bytes of data from this input stream.
+   *
+   * <p>Due to the "sparse" implementation of Large Objects, this class allows skipping
+   * past the "end" of the stream. Subsequent reads will continue to return {@code -1}.
```
Could you please explain what do you mean by "sparse" and what do you mean by "skipping past the end of the stream"?

The `skip(...)` contract asks to return the "actual number of bytes skipped". I am leaning towards that if we "skip past the end", we skip no bytes, so the return value should be `0`.

In other words, I would like to know what is the reason for `BlobInputStream#skip` implementation to mismatch `InputStream#skip` when considering "skip past the end".

davecramer replied:
Can you add the link to the comment so that future readers can find it?

OrangeDog replied:
This is probably a better link: https://www.postgresql.org/docs/14/lo-implementation.html
Unless there are objections regarding the version number?

davecramer replied:
you can always use https://www.postgresql.org/docs/current/lo-implementation.html

OrangeDog replied:
That tends to break when they move stuff around.

OrangeDog committed:
refactor: invalidate buffers before possible exception

OrangeDog committed:
refactor: defer initialisation of position to the first call that needs it

This prevents a surprise unchecked exception from the constructor.

OrangeDog commented:
Adding to the fun, CheckerFramework is definitely broken here. Even if I put `assert lo != null;` directly before the "offending" line, it still thinks it's an error.

vlsi commented:
> CheckerFramework is definitely broken here.

I am quite sure CheckerFramework is doing fine.

OrangeDog commented:
> I am quite sure CheckerFramework is doing the right thing.

OK then. How is `lo` possibly `null` at that line?

vlsi commented:
>OK then. How is lo possibly null at that line?

There are many methods in-between, and many of them could update `lo` value. See https://checkerframework.org/manual/#nullness-runtime-checks

OrangeDog commented:
> There are many methods in-between

No there aren't? Especially after I moved the null check to the immediately-preceding line.

davecramer commented:
> > There are many methods in-between
> 
> No there aren't? Especially after I moved the null check to the immediately-preceding line.

I'd be fine with suppressing the null check It is checked at the top of the function and an exception thrown so.

TBH. I've never completely understood why checker doesn't resolve these either

OrangeDog committed:
docs: link to details of how large objects are implemented

OrangeDog committed:
chore: tell CheckerFramework that value is non-null here

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,10 +346,88 @@ public boolean markSupported() {
+
   private LargeObject getLo() throws IOException {
-    if (lo == null) {
+    if (this.lo == null) {
       throw new IOException("BlobOutputStream is closed");
     }
+    LargeObject lo = castNonNull(this.lo);
```
Please remove the excessive `castNonNull`

OrangeDog replied:
[No.](https://github.com/pgjdbc/pgjdbc/pull/3148#issuecomment-2293242989)

vlsi commented on pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java:
```diff
@@ -335,10 +346,88 @@ public boolean markSupported() {
+      // Defer initialization until here so it can throw a checked exception
+      try {
+        // initialise current position for mark/reset
+        this.absolutePosition = lo.tell64();
+      } catch (SQLException e1) {
+        try {
+          // the tell64 function does not exist before PostgreSQL 9.3
+          this.absolutePosition = lo.tell();
```
This makes the behaviour slower for the database that do not support `tell64`. We can't accept such a change.

OrangeDog replied:
The last version that didn't support `tell64` was 9.2, which was EOL seven years ago. The annual PostgreSQL survey doesn't even split versions less than 10, so I cannot tell you how exactly few people that would affect.

As repeatedly explained, the only way to fix the bug is to make it slower. Though in real usage the addition time is unlikely to be noticeable compared to everything else the stream is doing.

If you have such a problem with "performance" then you should delete a load of code that prevents other bugs so that it can go a lot faster.

OrangeDog commented:
@davecramer I have had enough of these ridiculous comments from @vlsi. I will not be contributing any further.

This is the most hostile and incompetently-run open source project I have ever interacted with.

vlsi commented:
Thanks for this work, @OrangeDog. I've carried both fixes forward in #4204, keeping your commits with `Co-authored-by` and resolving #3144 and #3149. Closing this in favour of that PR. The groundwork is yours.
